FrEee/Scripts/AI/Vanguard/Plan.csx

```diff
@@ -0,0 +1,278 @@
+using FrEee.Utility;
+using System;
+using System.Collections.Generic;
+using FrEee.Utility.Extensions;
+using FrEee.Game.Objects.Orders;
+using FrEee.Game.Objects.Vehicles;
+using System.Linq; 
+
+public static class PlanManager
+{
+    public static Empire Empire;
+
+    public static PlanList CurrentPlans; 
+
+    /// <summary>
+    /// Unpacks all plans into 
+    /// </summary>
+    public static void Unpack()
+    {
+        if (Empire.AINotes.ContainsKey("PlanList"))
+        {
+            var str = Empire.AINotes["PlanList"] as string;
+            if (str != null)
+            {
+                CurrentPlans = FrEee.Utility.JsonSerializer.DeserializeObjectFromJson<PlanList>(str);
+                return; 
+            }
+        }          
+        CurrentPlans = new PlanList();
+        Empire.LogAIMessage("Failed to load plans"); 
+        
+    }
+
+    public static void Pack()
+    {
+        CurrentPlans.AssignAllIds(); 
+
+        if (Empire.AINotes.ContainsKey("PlanList"))
+            Empire.AINotes.Remove("PlanList");
+
+        var serializedPlans = FrEee.Utility.JsonSerializer.SerializeObjectToJson(CurrentPlans);
+        Empire.LogAIMessage(serializedPlans); 
+        Empire.AINotes.Add("PlanList", serializedPlans); 
+    }
+
+
+    public static void RemoveCompleted()
+    {
+        CurrentPlans.ColonizationPlans.RemoveAll(x => x.IsComplete);
+        CurrentPlans.ShipBuildPlans.RemoveAll(x => x.IsComplete); 
+    }
+
+    
+}
+
+/// <summary>
+/// The method of storing plans between turns for the AI. 
+/// </summary>
+[Serializable]
+public class PlanList
+{
+    public List<ColonizationPlan> ColonizationPlans;
+
+    public List<ShipBuildPlan> ShipBuildPlans; 
+
+    [DoNotSerialize]
+    public IEnumerable<Plan> AllPlans { 
+        get
+        {
+            var plans = new List<Plan>();
+            plans.AddRange(ColonizationPlans);
+            plans.AddRange(ShipBuildPlans);
+            return plans; 
+        } 
+    }
+
+
+    public PlanList()
+    {
+        ColonizationPlans = new List<ColonizationPlan>();
+        ShipBuildPlans = new List<ShipBuildPlan>(); 
+    }
+
+    public void AssignAllIds()
+    {
+        foreach (var p in ColonizationPlans)
+            if (p.PlanId == null)
+                p.AssignId();
+        foreach (var p in ShipBuildPlans)
+            if (p.PlanId == null)
+                p.AssignId();
+    }
+
+    /// <summary>
+    /// gets the chosen plan, or null. 
+    /// </summary>
+    /// <param name="planId"></param>
+    /// <returns></returns>
+    public Plan GetPlan(string planId)
+    {
+        return AllPlans.FirstOrDefault(x => x.PlanId == planId); 
+    }
+}
+
+
+[Serializable]
+public abstract class Plan
+{
+    public string PlanId; 
+
+
+    [DoNotSerialize]
+    public virtual string TypeShorthand { get; }
+
+    public bool IsComplete;
+
+    /// <summary>
+    /// The priority of the plan. Higher is better.  
+    /// </summary>
+    public int Priority; 
```

Comment:
It's kind of confusing that higher priority numbers are higher priority - usually "priority one" means top priority! If you want to use higher numbers to mean higher "priority", maybe call it something else?

Reply:
I mean, it sort of makes sense? The higher priority is the higher number. What do you suggest instead?

Reply:
I'm not sure, all I can think of is "urgency" but I don't really like that for some reason. As long as it's documented that higher numbers are "better" I guess we should be good.